AMD best practices

nadavgevaAMD opened the pull request:
After clearing the legal hurdles (for others within AMD as well), here is the AMD best practices validation layer. 

However, in the time it took to resolve, a necessary validation object (the command counter) has been removed and folded into CoreChecks. Before the PR can be accepted, I would like to discuss on how to best address adding the command counter to the validation layer. I would like to avoid duplicating the code to track the number and types of commands in a buffer, but could not see an obvious way to achieve this without restoring the command counter object. 

Another change I wanted to discuss is to the behavior of the best practice tests. Currently, the non-vendor tests enable all vendor checks as well. This triggers a large number of false errors, as the tests were not built with those checks in mind. It seems like going forward, enabling all vendor checks will create a large maintenance burden. With that in mind the non-vendor tests were changed to not enable vendor checks, and each vendor tests enables the vendor checks appropriate.

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=3144) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=3144) before we can accept your contribution.<br/><hr/>**Nadav Geva** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/Vulkan-ValidationLayers?pullRequest=3144) it.</sub>

jeremyg-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -19,6 +19,7 @@
  * Author: Mark Lobodzinski <mark@lunarg.com>
  * Author: Dave Houlton <daveh@lunarg.com>
  * Shannon McPherson <shannon@lunarg.com>
+ * Author: Nadav Geva <nadav.geva@amd.com>
```
It doesn't look like any code changes are added to these files in this patch, so Author lines are not needed.

nadavgevaAMD replied:
Thanks, will remove.

jeremyg-lunarg commented on layers/generated/command_validation.cpp:
```diff
@@ -626,12 +626,14 @@ static const std::array<const char *, CMD_RANGE_SIZE> kGeneratedBufferLevelList
     // in the ValidationObject object_dispatch. This is also making assumption this function is locked and this
     // write is not going to race without another copy of the CMD_BUFFER_STATE instance.
     const_cast<CMD_BUFFER_STATE *>(cb_state)->commandCount++;
+    const_cast<CMD_BUFFER_STATE *>(cb_state)->commands_in_buffer.push_back(cmd);
```
CMD_BUFFER_STATE::commands_in_buffer is not added to the class declaration in this patch, but it should be.

jeremyg-lunarg replied:
Also, I'm not entirely sure that this method gets called for the Best Practices validation object(s)....

nadavgevaAMD replied:
it is not, and one of the main points I wanted to discuss in the PR.

jeremyg-lunarg commented on layers/best_practices_validation.h:
```diff
@@ -147,6 +175,8 @@ class BestPractices : public ValidationStateTracker {
 
     std::string GetAPIVersionName(uint32_t version) const;
 
+    void InitDeviceValidationObject(bool add_obj, ValidationObject* inst_obj, ValidationObject* dev_obj);
```
This should be marked with the override keyword. Probably clang will complain on the Android build without it.

nadavgevaAMD replied:
fixed

jeremyg-lunarg commented on layers/best_practices_validation.h:
```diff
@@ -472,6 +502,69 @@ class BestPractices : public ValidationStateTracker {
+                                                             const VkAllocationCallbacks* pAllocator, VkPipeline* pPipelines,
+                                                             VkResult result, void* state_data);
+
+    void PostCallRecordCmdPipelineBarrier(VkCommandBuffer commandBuffer, VkPipelineStageFlags srcStageMask,
```
All of the Pre/Post Record and Validate functions also probably need the 'override' keyword.

nadavgevaAMD replied:
should this be moved to `ManualPostCallRecordCmdPipelineBarrier`?

at one point I understood the logic of when to make a Manual version of these function, but it escapes me now...

nadavgevaAMD replied:
Ah yes, I remember now. I've fixed all the Pre/Post functions with override.

jeremyg-lunarg commented on layers/cmd_buffer_state.h:
```diff
@@ -189,6 +191,7 @@ class CMD_BUFFER_STATE : public REFCOUNTED_NODE {
     bool unprotected;  // can't be used for protected memory
 
     CB_STATE state;         // Track cmd buffer update state
+    std::vector<CMD_TYPE> commands_in_buffer; // the command types in the buffer, used in the best practice layer
```
This belongs in an earlier patch.

nadavgevaAMD replied:
will move it, but see earlier question on how to best implement this

jeremyg-lunarg commented on scripts/best_practices_generator.py:
```diff
@@ -89,12 +90,12 @@ def __init__(self,
         OutputGenerator.__init__(self, errFile, warnFile, diagFile)
         # Commands which are not autogenerated but still intercepted
-        self.no_autogen_list = [
+        self.no_autogen_list = [            
```
It looks like there are whitespace errors here.

nadavgevaAMD replied:
removed

jeremyg-lunarg commented on tests/vklayertests_arm_best_practices.cpp:
```diff
@@ -20,7 +22,7 @@
     TEST_DESCRIPTION("Test for multisampled images with too many samples");
 
-    InitBestPracticesFramework();
+    InitBestPracticesFramework("VALIDATION_CHECK_ENABLE_VENDOR_SPECIFIC_ARM");
```
It would be helpful to define a const for this string to avoid repeating the literal.

nadavgevaAMD replied:
moved to
```
#define ENABLE_ARM_VALIDATION "VALIDATION_CHECK_ENABLE_VENDOR_SPECIFIC_ARM"
```
at the head of the file.

ncesario-lunarg replied:
Would be great to have that be a `constexpr char *`, or something to that effect, just not a macro, please :)

nadavgevaAMD replied:
changed to 
```
constexpr char * ENABLE_ARM_VALIDATION = "VALIDATION_CHECK_ENABLE_VENDOR_SPECIFIC_ARM";
```

nadavgevaAMD replied:
This is the only `constexpr char` in the codebase. Any preference for the style of the variable name?

ncesario-lunarg replied:
I think most of our string constants are declared`static const char *`, though that was probably before the move to C++11. I think we're probably supposed to be using `constexpr` based on https://google.github.io/styleguide/cppguide.html#Constant_Names, but either is fine with me.

Constant names should have a lower case `k` prefix followed by CamelCase, so this should probably be `kEnableArmValidation`: https://google.github.io/styleguide/cppguide.html#Constant_Names

ncesario-lunarg replied:
Unfortunately, we don't have consistent code styling across all files (yet...using clang-tidy to get closer is on the todo list :)). For now, we try to follow the Google C++ style guide for new code that's added, or update to the appropriate style when code is modified.

nadavgevaAMD replied:
Changed to `kEnableArmValidation` and `kEnableAMDValidation`

jeremyg-lunarg reviewed (changes requested):
Thank you for the patches!  I have a lot of little comments but one big concern: I am worried that adding the 'track every command in the command buffer' code to CMD_BUFFER_STATE may cause performance problems in Core validation.  We may need to do something similar to what @ncesario-lunarg  did with class SWAPCHAIN_STATE_BP in [this change](https://github.com/KhronosGroup/Vulkan-ValidationLayers/commit/39152e64829c35cdad6a6d667cf70428b39e3f6f#diff-b63daf1ca90346ea775b2a8de4ee9e40a34ae940fb91593d7602f2eccb9e92df).  I'm working on some cleanups in to CMD_BUFFER_STATE and can probably add some stuff to make this easier to do. I'll add you to my PR when it goes up, which will probably be this week.

christophe-lunarg reviewed (changes requested):
The validation layer manifest needs to be updated to expose the AMD Best Practices setting to the users using vkconfig and generated documentation.

By the way: Exciting work!

nadavgevaAMD commented on layers/layer_options.cpp:
```diff
@@ -1,6 +1,7 @@
-/* Copyright (c) 2020-2021 The Khronos Group Inc.
- * Copyright (c) 2020-2021 Valve Corporation
- * Copyright (c) 2020-2021 LunarG, Inc.
+/* Copyright (c) 2021 The Khronos Group Inc.
```
Looks like it reverted to an earlier version of the copyright after rebasing. I'll fix it.

nadavgevaAMD commented on scripts/layer_chassis_generator.py:
```diff
@@ -657,6 +660,7 @@ class ValidationObject {
  * limitations under the License.
  *
  * Author: Mark Lobodzinski <mark@lunarg.com>
+ * Author: Nadav Geva <nadav.geva@amd.com>
```
this came from lawyers, to add this to every file I touched :\

jeremyg-lunarg replied:
Ok, that's fine then.

nadavgevaAMD commented:
general question on addressing the comments: would you prefer a single "this commit address code review comments" or a rebase with a "clean" history as if everything occured originally?

ncesario-lunarg commented on layers/best_practices_utils.cpp:
```diff
@@ -3162,6 +3396,316 @@ bool BestPractices::PreCallValidateCreateSampler(VkDevice device, const VkSample
+    ValidationStateTracker::PreCallRecordCreateGraphicsPipelines(device, pipelineCache, createInfoCount, pCreateInfos, pAllocator,
+                                                                 pPipelines);
+    // AMD best practice
+    num_pso += createInfoCount;
```
Does this need to be adjusted at `DestroyPipeline` time?

ncesario-lunarg replied:
Should this happen in the `Post` record function, rather than the `Pre`, so that we know whether or not the dispatch call succeeded before updating the state?

nadavgevaAMD replied:
The performance concern here is more about the general numbers and not the current number of active pipelines. Creating and destroying 1 pipeline repeatedly, creating many pipelines that failed (assuming this was missed by other validations), or having just many pipelines in general, are all bad behaviors to catch and alert the user.

ncesario-lunarg commented on layers/best_practices_utils.cpp:
```diff
@@ -3162,6 +3396,316 @@ bool BestPractices::PreCallValidateCreateSampler(VkDevice device, const VkSample
+        if (pipeline_size > kPipelineLayoutSizeWarningLimit) {
+            skip |= LogPerformanceWarning(device, kVUID_BestPractices_CreatePipelinesLayout_KeepLayoutSmall,
+                        "%s Performance warning: pipeline layout size is too large. Prefer smaller pipeline layouts"
+                        "Descriptor sets cost 1 DWORD each.                                   "
```
This might look a bit odd when logging with all of the spaces.

nadavgevaAMD replied:
Good catch, looks like some auto-format gone wrong. Fixed

ncesario-lunarg reviewed (changes requested):
Looks like a lot of good info!

I didn't quite make it through the whole thing, but I think @jeremyg-lunarg's comments are more important to get addressed first.

ncesario-lunarg commented:
> general question on addressing the comments: would you prefer a single "this commit address code review comments" or a rebase with a "clean" history as if everything occured originally?

It it's not too difficult, a rebase with a "clean" history would be preferred, at least before the change gets merged. That helps make things a bit easier to understand when going back through the history, especially when bisecting.

nadavgevaAMD commented on layers/best_practices_error_enums.h:
```diff
@@ -2,6 +2,7 @@
  * Copyright (c) 2015-2021 Valve Corporation
  * Copyright (c) 2015-2021 LunarG, Inc.
  * Copyright (C) 2015-2021 Google Inc.
+ * Modifications Copyright (C) 2020 Advanced Micro Devices, Inc. All rights reserved.
```
Hmm I'm not sure. I'll double check with the lawyers.

nadavgevaAMD replied:
@Tobski confirmed that this is just the copyright message AMD lawyers want. All AMD copyright notifications in the validation layers (including outside of this PR) contain this message.

nadavgevaAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -306,6 +320,39 @@ bool BestPractices::PreCallValidateCreateImage(VkDevice device, const VkImageCre
+        if ((pCreateInfo->usage & (VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT)) &&
+            (pCreateInfo->sharingMode == VK_SHARING_MODE_CONCURRENT)) {
+            skip |= LogPerformanceWarning(device,
+                            kVUID_BestPractices_vkImage_AvoidConcurentRT,
```
Fair point. This started before "RT" was so confusing. Changed to `kVUID_BestPractices_vkImage_AvoidConcurentRenderTargets`

nadavgevaAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -306,6 +320,39 @@ bool BestPractices::PreCallValidateCreateImage(VkDevice device, const VkImageCre
+                        VendorSpecificTag(kBPVendorAMD), image_hex.str().c_str());
+        }
+        
+        if ((pCreateInfo->usage & VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT) && (pCreateInfo->usage & VK_IMAGE_USAGE_STORAGE_BIT) ||
+            (pCreateInfo->usage & VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT) &&
```
Not quite, since we are checking for Color attachment and Storage, or Depth and Storage. However, changed to match the style better:
```
        if ((pCreateInfo->usage & (VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT)) &&
                (pCreateInfo->usage & VK_IMAGE_USAGE_STORAGE_BIT)) {
```

ncesario-lunarg replied:
:+1: missed the `||` in there.

nadavgevaAMD commented on layers/best_practices_validation.h:
```diff
@@ -497,6 +591,16 @@ class BestPractices : public ValidationStateTracker {
+    uint32_t num_queue_submissions = 0;
+    VkPipelineCache pipeline_cache = 0;
+    layer_data::unordered_set<VkPipeline> pipelines_used_in_frame;
+    mutable bool robust_buffer_access = false;
```
Good catch, looks like it got lost in a rebase. I'll fix it. It was mutable because the best place to write it was in an initialization const function. Maybe whatever caused it to be removed also introduced a better way

nadavgevaAMD replied:
Ok, found where it was. I've added it in `PreCallValidateCreateDevice` since there was already a check there (by Arm) for robust buffer access. I didn't want to introduce a `PostCallRecordCreateDevice` function just for this one flag, but that `PreCall` is const function.  I can move it to a `PostCallRecord` if that's better

ncesario-lunarg replied:
@Tony-LunarG @jeremyg-lunarg Any thoughts?

I know adding another call is annoying, but I'd also rather not add `mutable` variables unless absolutely necessary (otherwise having a function marked `const` becomes less meaningful, or at least harder to reason about what's changed).

nadavgevaAMD commented on layers/best_practices_validation.h:
```diff
@@ -38,6 +40,34 @@ static const VkDeviceSize kMinDeviceAllocationSize = 256 * 1024;
+// Note: These are initial ball park numbers for good performance
+// We expect to adjust them as we get more data on layer usage
+// Avoid small command buffers
+static const uint32_t kMinRecommendedCommandBufferSize = 10;
```
Changed all constants in this list to end with `AMD`

nadavgevaAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -1331,6 +1431,60 @@ bool BestPractices::PreCallValidateCmdPipelineBarrier(VkCommandBuffer commandBuf
+
+            // general no with storage
+            if (pImageMemoryBarriers[i].newLayout == VK_IMAGE_LAYOUT_GENERAL) {
+                auto image_state = GetImageState(pImageMemoryBarriers[i].image);
```
Changed all instances of `GetImageState` to `Get<IMAGE_STATE>`

nadavgevaAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -3162,6 +3396,316 @@ bool BestPractices::PreCallValidateCreateSampler(VkDevice device, const VkSample
+                num_pipeline_bind_cmds++;
+            }
+        }
+        if (num_draw_cmds * 1.0 / num_pipeline_bind_cmds < kDrawsPerPipelineRatioWarningLimit) {
```
changed to:
`if (num_draw_cmds * (1.0 / num_pipeline_bind_cmds) < kDrawsPerPipelineRatioWarningLimitAMD) {`

nadavgevaAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -3162,6 +3396,316 @@ bool BestPractices::PreCallValidateCreateSampler(VkDevice device, const VkSample
+        if (num_stream_bind_cmds * 1.0 / num_draw_cmds > kVertexStreamToDrawRatioWarningLimit) {
+            skip |= LogPerformanceWarning(device, kVUID_BestPractices_DrawState_AvoidVertexBindEveryDraw,
+                        "%s Performance warning: In command buffer %s, the number of vkCmdBindVertexBuffers is similar to the number "
+                        "of draw commands"
```
fixed

nadavgevaAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -1331,6 +1431,60 @@ bool BestPractices::PreCallValidateCmdPipelineBarrier(VkCommandBuffer commandBuf
+    if (VendorCheckEnabled(kBPVendorAMD)) {
+        if (num_barriers_objects + imageMemoryBarrierCount + bufferMemoryBarrierCount > kMaxRecommendedBarriersSize) {
+            skip |= LogPerformanceWarning(device, kVUID_BestPractices_CmdBuffer_highBarrierCount,
+                        "%s Performance warning: In this frame, %d barriers were already submitted. Barriers have a high cost and can "
```
Changed. Is there an equivalent for `%s`?

ncesario-lunarg replied:
I don't think so. The big thing is to add some semblance of "type safety" w.r.t. printf parameters. AFAIK there is only one type of C-string.

christophe-lunarg commented:
I made a PR in Vulkan Configurator so that the AMD best practices get exposed by the UI when the validation layer manifest has it:
https://github.com/LunarG/VulkanTools/pull/1587

nanokatze commented on layers/best_practices_utils.cpp:
```diff
@@ -949,6 +1004,33 @@ bool BestPractices::PreCallValidateCreateGraphicsPipelines(VkDevice device, VkPi
+        }
+
+        // TODO: this might be too aggressive of a check
+        if (pCreateInfos->pDynamicState && pCreateInfos->pDynamicState->dynamicStateCount > kDynamicStatesWarningLimit) {
```
This appears rather arbitrary and of questionable utility. Taking a look at AMDVLK we find that the source is a little hard to read and navigate, so instead, with RADV as our next best reference, we find that the only dynamic states that do end up being an obvious penalty are DEPTH_TEST_ENABLE_EXT, DEPTH_WRITE_ENABLE_EXT, DEPTH_COMPARE_OP_EXT, STENCIL_TEST_ENABLE_EXT and STENCIL_OP_EXT, causing "out of order rasterization" to be disabled [^1]. It seems fair to expect AMDVLK to follow similar pattern, being absolutely oblivious to dynamicity of some states, while requiring more time and/or space for processing (underline as appropriate) for other dynamic states, which is what one would expect this code to be checking. Would it be possible for this check to be replaced with more precise ones and for the messages to be more detailed?

[^1]: https://gitlab.freedesktop.org/mesa/mesa/-/blob/a1b7a5fad6c6773f7ee71844eb92d000f098e797/src/amd/vulkan/radv_pipeline.c#L894

gevahn replied:
This one was tricky. Dynamic states have diffuse costs. There is the cost of vkCmdSet* commands, a small cost at pipeline creation time, a small cost at pipeline bind time, etc. In isolation, these are fine. But we've seen cases where people set all states as dynamic in the pipeline, and just set default values. It was this kind of API misuse this check was for. As the TODO states, we expected to tweak this value/check as we get more information from the userbase.

nadavgevaAMD replied:
(accidentally responded from my personal github, sorry!)

gevahn commented on layers/best_practices_error_enums.h:
```diff
@@ -179,4 +181,62 @@ static const char DECORATE_UNUSED *kVUID_BestPractices_RenderPass_CopyImage_Load
+    "UNASSIGNED-BestPractices-vkImage-DontUseMutableRT";
+static const char DECORATE_UNUSED *kVUID_BestPractices_vkImage_AvoidImageToImageCopy =
+    "UNASSIGNED-BestPractices-vkImage-AvoidImageToImageCopy";
+static const char DECORATE_UNUSED *kVUID_BestPractices_vkImage_AvoidConcurentRT =
+    "UNASSIGNED-BestPractices-vkImage-AvoidConcurentRT";
```
Good catch, changed

jeremyg-lunarg commented:
> ... We may need to do something similar to what @ncesario-lunarg did with class SWAPCHAIN_STATE_BP in [this change](https://github.com/KhronosGroup/Vulkan-ValidationLayers/commit/39152e64829c35cdad6a6d667cf70428b39e3f6f#diff-b63daf1ca90346ea775b2a8de4ee9e40a34ae940fb91593d7602f2eccb9e92df). I'm working on some cleanups in to CMD_BUFFER_STATE and can probably add some stuff to make this easier to do. I'll add you to my PR when it goes up, which will probably be this week.

Based on how much code I'm changing, the work above is going to be at least 3 PRs to avoid breaking reviewers brains...
1.  https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3149 - makes CMD_BUFFER_STATE more OO-y
2. TODO - convert existing layers to subclass CMD_BUFFER_STATE  like SWAPCHAIN_STATE_BP above. I have this mostly working now.
3. TODO - add a method to CMD_BUFFER_STATE that works kinda like ValidateCmd() but happens in RecordCmd* methods.  Then best practices can override this method to record cmd types like in this patch.

nadavgevaAMD commented:
I've addressed all the smaller comments. I'm preparing a clean rebase with them, and will organise it such that I can add the new cmd_buffer_state functionality when it comes online. There also seems to be a build issue on the CI? I can build locally, so must have missed something

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

Tony-LunarG commented:
I ran this through CTS last night and no crashes

Tony-LunarG commented:
> I'm trying to resolve the CI errors, but I'm not sure where are they coming from. In particular :
> 
> ```
> ERROR: Repo files do not match generator output for chassis_dispatch_helper.h
> ```
> 
> I'm not touching this file or it's generator anymore in this PR. Could it be a cached version or something?

You added yourself as author in the generator script, but didn't check in layers/generated/chassis.cpp or layers/generated/chassis_dispatch_helper.h that resulted from running the generator (adding you as author)

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 18689.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 4377 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 4377 failed.

jeremyg-lunarg commented on scripts/best_practices_generator.py:
```diff
@@ -94,7 +95,7 @@ def __init__(self,
             'vkDestroyValidationCacheEXT',
             'vkMergeValidationCachesEXT',
-            'vkGetValidationCacheDataEXT',
+            'vkGetValidationCacheDataEXT',            
```
still one whitespace error here.

nadavgevaAMD replied:
they always sneak past me... fixed

jeremyg-lunarg reviewed (commented):
It looks to me like this is getting really close, except for the stuff that requires the commands_in_buffer.  Would you consider pulling that part out into a separate PR? Then the rest could probably go in sooner.

nadavgevaAMD commented:
I can pull the commands_in_buffer out to a separate commit, it's not too tangled. I can do that Monday, and the rest can go in if deemed ready. I do want to come back to the checks that need commands_in_buffer quickly, since they are some of the most important checks for us. Looks like the PR for the CMD_BUF_STATE will everything I need easily.

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

nadavgevaAMD commented:
I've pulled out the checks and tests that needed CB_STATE. I think I did it correctly, but a once-over to check that I didn't miss anything is probably needed. Otherwise, it is ready to go in.

jeremyg-lunarg reviewed (approved):
All my concerns are addressed, thank you! I hope to have the last PR you'll need for the command counting part posted later today.

@ncesario-lunarg & @christophe-lunarg , what do you think?

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ncesario-lunarg commented on tests/vklayertests_amd_best_practices.cpp:
```diff
@@ -0,0 +1,769 @@
+        // also imitates a "just in time" pipeline creator pattern
+        if (i == 1) {
+            // check that the second pipeline helper cache was detected
+            m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT,
```
You should be able to use `kPerformanceWarningBit` for brevity if you prefer.

nadavgevaAMD replied:
Thanks! Will switch to that.

ncesario-lunarg commented on tests/layer_validation_tests.h:
```diff
@@ -290,6 +290,7 @@ class VkPositiveLayerTest : public VkLayerTest {
 class VkBestPracticesLayerTest : public VkLayerTest {
   public:
     void InitBestPracticesFramework();
+    void InitBestPracticesFramework(const char* ValidationChecksToEnable);
```
`snake_case` for variables please.

nadavgevaAMD replied:
fixed

nadavgevaAMD replied:
fixed

ncesario-lunarg reviewed (changes requested):
Other than some very minor nits, looks good!

nadavgevaAMD commented on docs/best_practices.md:
````diff
@@ -54,6 +54,42 @@ VK_LAYER_ENABLES=VK_VALIDATION_FEATURE_ENABLE_BEST_PRACTICES_EXT;VALIDATION_CHEC
 VK_LAYER_ENABLES=VK_VALIDATION_FEATURE_ENABLE_BEST_PRACTICES_EXT:VALIDATION_CHECK_ENABLE_VENDOR_SPECIFIC_ARM
 ```
 
+Similarly to enable the AMD-specific best pratices checks, us the following:
````
thanks, fixed

nadavgevaAMD commented on layers/generated/best_practices.h:
```diff
@@ -21,6 +21,7 @@
  * limitations under the License.
  *
  * Author: Mark Lobodzinski <mark@lunarg.com>
+ * Author: Nadav Geva <nadav.geva@amd.com>
```
For this (and the ones below) I was asked to add this to all files I touched. However, these are so minor I'm going to remove it. Certainly, I wouldn't want to be contacted if there is an issue in these files...

ncesario-lunarg replied:
There's...nothing changed in these files, correct? Or am I missing something?

jeremyg-lunarg replied:
This is generated  with the same code generation header template as the other files that were changed, so we're kinda stuck with these getting added.

ncesario-lunarg replied:
Ah, right. Thanks for the clarification @jeremyg-lunarg. That is unfortunate.

nadavgevaAMD replied:
Oh these are because the cpp file has additions, but the generator adds the same copyright message to both the header and implementation.

nadavgevaAMD commented on tests/vklayertests_amd_best_practices.cpp:
```diff
@@ -0,0 +1,769 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Author: Nathaniel Cesario <nathaniel@lunarg.com>
```
Indeed, this file started as a copy of the arm version (for which you are the sole author listed). However, I don't think there is much overlap beyond the usual bits and pieces of initialization code. I can be sole author on this if you prefer.

nadavgevaAMD commented on tests/vklayertests_amd_best_practices.cpp:
```diff
@@ -0,0 +1,769 @@
+    // disabled for now
+    return;
+
+InitBestPracticesFramework(kEnableAMDValidation);
```
I think so, if I understand the CI log correctly.

ncesario-lunarg replied:
Right, I was just wondering if it ran locally. It looks like the indentation is incorrect here. If that is the case, maybe something is wrong with clang-format; or I'm seeing things :)

nadavgevaAMD replied:
Oh you are right, it is not indented correctly! maybe the return confused clang-format. fixed!

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

nadavgevaAMD commented:
Fixed the nits, and some other complaints gcc and clang had. I'm too used to MSVC... 
If this clears CI, should be good to go.

ncesario-lunarg reviewed (approved):
LGTM, thanks for all the quick fixes!

ncesario-lunarg commented:
So close :) https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3144/checks?check_run_id=3365950871#step:5:3014

Looks like this is happening on RADV as well as the mock ICD.

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

nadavgevaAMD commented:
Foiled by my own tests! That's what I get for trying to rush things.... Should be fixed with latest change.

ci-tester-lunarg commented:
Author nadavgevaAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 22423.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 4469 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 4469 failed.

nadavgevaAMD commented:
The CI failed, but all the checks passed? Is there a log somewhere with the reason?

ncesario-lunarg commented:
You should get an automated email containing logs; I'll see if something is either wrong or needs additional setup on our end.

## Windows failures
Unfortunately we support pre-VS2019 compilers and we have internal CI that runs to verify these builds, so this is one of the errors we're getting (which won't show up in VS2019):
```
descriptor_sets.cpp(356): warning C4800: 'VkDescriptorSetLayoutCreateFlags': forcing value to bool 'true' or 'false' (performance warning)
```

## Android failures
Android built, but crashed on `VkLayerTest.InvalidExportExternalImageHandleType`; this seems coincidental at this point.

## Linux failures
Our linux builds are failing with:
```
vklayertests_amd_best_practices.cpp:39:16: error: variable pipeline_Array set but not used
```

And then our RADV setup is crashing on `VkLayerTest.CreateDescriptorSetLayoutWithInvalidFlags`.

ncesario-lunarg commented:
We're working to get these failures to show up in the list of github checks, but we haven't been able to set that up yet.

nadavgevaAMD commented:
The only email I received just said "build failed", with no log...

These errors are strange. Maybe need to rebase? I don't have `VkDescriptorSetLayoutCreateFlags` at all in this PR. Similarly, `VkLayerTest.CreateDescriptorSetLayoutWithInvalidFlags` is unrelated to anything I've done.

`pipeline_Array` is part of the disabled long test for excessive pipeline creation. I can comment out the test, or remove it.
